File: app/src/main/java/com/example/globoplay_desafio_mobile/ui/watchlist/WatchListFragment.java
package com.example.globoplay_desafio_mobile.ui.watchlist;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import com.example.globoplay_desafio_mobile.R;
import com.example.globoplay_desafio_mobile.adapters.GridViewAdapter;
import com.example.globoplay_desafio_mobile.databinding.WatchlistFragmentBinding;
import com.example.globoplay_desafio_mobile.interfaces.OnMovieListener;
import com.example.globoplay_desafio_mobile.models.movies.MovieModel;
import com.example.globoplay_desafio_mobile.utils.Variables;
import com.example.globoplay_desafio_mobile.viewmodels.watchlist.WatchListViewModel;

import java.util.List;

import static android.content.Context.MODE_PRIVATE;

public class WatchListFragment extends Fragment implements OnMovieListener {

    //ViewBinding
    private WatchlistFragmentBinding watchlistFragmentBinding;
    //Adapter
    private GridViewAdapter gridViewAdapter;
    //ViewModel
    WatchListViewModel watchListViewModel;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        watchlistFragmentBinding = WatchlistFragmentBinding.inflate(inflater,container,false);

        watchListViewModel = new ViewModelProvider(requireActivity()).get(WatchListViewModel.class);

        Variables.WATCHLIST = true;

        //ConfigureGridView
        configureGridView();

        //Calling the observer
        observerWatchListChange();

        //get info user
        SharedPreferences prefs = getActivity().getSharedPreferences("detailsUser", MODE_PRIVATE);
        String sessionId = prefs.getString("sessionId","");
        String accountId = prefs.getString("accountId","");

        if(gridViewAdapter.movies !=null) {
            if(gridViewAdapter.movies.size() == 0) {
                //get WatchList
                watchListViewModel.getWatchList(accountId,sessionId);
            }
        }
        else {
            //get WatchList
            watchListViewModel.getWatchList(accountId,sessionId);
        }


        return watchlistFragmentBinding.getRoot();
    }

    //observer
    private void observerWatchListChange() {
        watchListViewModel.getMovies().observe(getViewLifecycleOwner(), new Observer<List<MovieModel>>() {
            @Override
            public void onChanged(List<MovieModel> movieModels) {
                Variables.watchList = movieModels;
                gridViewAdapter.movies = movieModels;
                gridViewAdapter.notifyDataSetChanged();
            }
        });
    }


    //configure GridView
    private void configureGridView() {
        gridViewAdapter = new GridViewAdapter(this);
        watchlistFragmentBinding.gridview.setAdapter(gridViewAdapter);
    }

    //when user click on the movie
    @Override
    public void onMovieClick(MovieModel movie) {
        Bundle bundle = new Bundle();
        bundle.putParcelable("movie", movie);
        Navigation.findNavController(getActivity(), R.id.nav_host_fragment).navigate(R.id.movieDetailsFragment,bundle);
    }

}

File: app/src/main/java/me/davidpcosta/tmdb/ui/main/watchlist/WatchlistFragment.kt
package me.davidpcosta.tmdb.ui.main.watchlist

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.GridView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import me.davidpcosta.tmdb.R
import me.davidpcosta.tmdb.data.model.Movie
import me.davidpcosta.tmdb.toast
import me.davidpcosta.tmdb.ui.highlight.HighlightActivity


class WatchlistFragment : Fragment() {

    private lateinit var watchlistViewModel: WatchlistViewModel
    private lateinit var watchlistGrid: GridView
    private lateinit var movieAdapter: MovieAdapter
    private lateinit var sharedPreferences: SharedPreferences

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.activity_main_fragment_watchlist, container, false)
        sharedPreferences = requireActivity().getSharedPreferences(getString(R.string.const_shared_preference), Context.MODE_PRIVATE)
        val sessionId = sharedPreferences.getString(getString(R.string.const_key_session_id), "")!!
        val accountId = sharedPreferences.getLong(getString(R.string.const_key_account_id), 0)

        watchlistViewModel = ViewModelProvider(this, WatchlistViewModelFactory(requireActivity().applicationContext)).get(WatchlistViewModel::class.java)
        movieAdapter =
            MovieAdapter(requireActivity().applicationContext)

        watchlistGrid = view.findViewById<GridView>(R.id.watchlist).apply {
            adapter = movieAdapter
            onItemClickListener =  AdapterView.OnItemClickListener { _, _, position, _ ->
                val movie = watchlistViewModel.movies.value!![position]
                requireActivity().toast(movie.title)
                goToMovie(movie)
            }
        }

        watchlistViewModel.fetchWatchlist(accountId, sessionId)
        watchlistViewModel.movies.observe(viewLifecycleOwner, Observer {
            movieAdapter.movies = it
            movieAdapter.notifyDataSetChanged()
        })

        return view
    }

    private fun goToMovie(movie: Movie) {
        val intent = Intent(requireActivity(), HighlightActivity::class.java)
        intent.putExtra(getString(R.string.const_key_movie), movie)
        requireActivity().startActivity(intent)
    }
}
